composer/loggers/slack_logger.py

```diff
@@ -0,0 +1,190 @@
+2. SLACK_LOGGING_CHANNEL_ID: Channel id to send the message (Open slack channel in web browser to look this up)
+
+Next write script to output metrics / hparams / traces to slack channel. See example below:
+
+trainer = Trainer(
+    model=mnist_model(num_classes=10),
+    train_dataloader=train_dataloader,
+    max_duration='2ep',
+    algorithms=[
+        LabelSmoothing(smoothing=0.1),
+        CutMix(alpha=1.0),
+        ChannelsLast(),
+    ],
+    loggers=[
+        SlackLogger(
+            log_metrics_config=SlackLogConfig(
+                formatter_func=(lambda data, **kwargs: [{
+                    'type': 'section',
+                    'text': {
+                        'type': 'mrkdwn',
+                        'text': f'*{k}:* {v}'
+                    }
+                } for k, v in data.items()]),
+                include_keys={'loss/train/total'},
+                interval_in_seconds=1
+            ),
+            log_hparams_config=SlackLogConfig(),
+        )
+    ],
+)
+
+trainer.fit()
+"""
+
+import os
+import time
+from typing import Any, Callable, Dict, List, Optional, Set, Tuple
+
+from composer.loggers.logger_destination import LoggerDestination
+from composer.utils import MissingConditionalImportError
+
+
+class SlackLogConfig:
+    """Log config for slack logger, including formatter function, keys to include, and interval in seconds.
+
+    formatter_func ((...) -> Any | None): A formatter function that returns list of blocks to be sent to slack
+    include_keys (Set[str]): A set of metric/logs/traces keys to include in the message. If None, no keys are included.
+    interval_in_seconds (int): Interval in seconds to send logs to slack. If None, message is sent after every log call.
+    Note that no more than 50 items are allowed to send in a single message.
+    """
+
+    def __init__(
+        self,
+        formatter_func: Optional[Callable[..., List[Dict[str, Any]]]] = None,
+        include_keys: Optional[Set[str]] = None,
+        interval_in_seconds: Optional[int] = None,
+    ) -> None:
+        self.formatter_func = formatter_func
+        self.include_keys = include_keys
+        self.interval_in_seconds = interval_in_seconds
+
+
+class SlackLogger(LoggerDestination):
+    """Logger to log metrics to slack.
+
+    This logger uses Slack's postMessage api - https://api.slack.com/methods/chat.postMessage.
+
+    Args:
+        log_metrics_config (SlackLogConfig): Configuration for logging metrics to slack.
+        log_traces_config (SlackLogConfig): Configuration for logging traces to slack.
+        log_hparams_config (SlackLogConfig): Configuration for logging hyperparameters to slack.
+    """
+
+    def __init__(
+        self,
+        log_metrics_config: Optional[SlackLogConfig] = None,
+        log_traces_config: Optional[SlackLogConfig] = None,
+        log_hparams_config: Optional[SlackLogConfig] = None,
```

Comment:
```suggestion
        formatter_func: Optional[Callable[..., List[Dict[str, Any]]]] = None,,
        include_keys: Optional[Set[str]] = None,
        interval_in_seconds: Optional[int] = None,
```

I think having the SlackLogConfig for each category logged is a bit of overkill:

1. It is kinda an unnecessary lift for a user learn about this new class and instantiate a new object for every type of filtering they want to do.
2. It is probably overkill to have a different config for traces, hyperparameters, and metrics. (Traces are only really used as debugging and it is easy for a user to mistake certain things as hyperparameters, which are actually metrics, so the user would be playing a game of guess and check to set the right config.
3. Seems like we want just an MVP. This seems like a premature abstraction

We could also a do a dictionary instead of flat arguments

Reply:
yes that makes sense, updated!